dashboard/app/models/level_source.rb

```diff
@@ -51,4 +51,26 @@ def self.find_identical_or_create(level, data)
       end
     end
   end
+
+  # @param [Integer] user_id The ID of the user performing the obfuscation.
+  # @return [String] The obfuscated (with the user_id) level source ID.
+  def obfuscate_level_source_id(user_id)
+    Base64.urlsafe_encode64("#{id}:#{user_id}")
+  end
+
+  # @param [String] obfuscated_level_source_id_user_id The obfuscated (with the user_id)
+  #   level_source ID.
+  # @param [Boolean] ignore_missing_user Whether to ignore the absence of the user specified by the
+  #   user ID when returning the level source ID. Default false.
+  # @return [Integer, nil] The (deobfuscated) level source ID, returns nil if the user specified in
+  #   the obfuscation does not exist.
+  def self.deobfuscate_level_source_id(obfuscated_level_source_id_user_id, ignore_missing_user: false)
+    level_source_id, user_id = Base64.urlsafe_decode64(obfuscated_level_source_id_user_id).split(':')
+    if ignore_missing_user || user_id.nil? || User.find_by_id(user_id)
```

Comment:
There is no user_id if a non-logged-in user makes the levelsource. This is not super secure, but my understanding is that it's not meant to be. It's more of a way to not explicitly put a user_id in the URL. Also level_source is not specific to specific users so really we just want that specific user's url to not work if they are deleted. People will still be able to see the code if another user made it and shared it.

@ashercodeorg can correct me if I'm wrong.